# Lab 3

In [28]:
from classes import PatientExam
from main2 import load_exams, average_bmi, busiest_month

## Step 1:
Each exam record becomes an object with 5 attributes (one per CSV column), with two methods: `calculate_bmi()` and `get_month()`.

Creating an exam and checking its attributes

In [29]:
exam = PatientExam(1, "12/17/2002", "John Smith", 90, 1.76)
print(exam.exam_id)
print(exam.date)
print(exam.patient_name)
print(exam.weight)
print(exam.height)

1
12/17/2002
John Smith
90
1.76


**calculate_bmi()** = weight (kg) ÷ height (m)².

In [ ]:
bmi = exam.calculate_bmi()
print(bmi)

print(90 / (1.76 ** 2))

29.054752066115704
29.054752066115704


**get_month():** Since the date is stored as a string like `"12/17/2002"` (month/day/year). Rather than parsing a the full object, this just splits on `/` and takes the first piece.

In [31]:
print("12/17/2002".split("/"))
month = exam.get_month()
print(month)

['12', '17', '2002']
12


__repr__ :  makes sure the patientexam prints out readable

In [32]:
print(exam)

PatientExam(1, 12/17/2002, John Smith)


## Step 2: Loading the dataset
load_exams reads patient_data.csv, strips the header and builds per line, converting weight and height to floats and the exam id to an int.

In [33]:
exams = load_exams("patient_data.csv")
print(len(exams))
print(exams[0])
print(exams[-1])

1000
PatientExam(1, 12/17/2002, Wyn Grey)
PatientExam(1000, 10/14/2002, Aeris Solis)


Checking a few individual attributes to confirm types converted correctly:

In [34]:
print(type(exams[0].exam_id), exams[0].exam_id)
print(type(exams[0].weight), exams[0].weight)
print(type(exams[0].height), exams[0].height)

<class 'int'> 1
<class 'float'> 90.0
<class 'float'> 1.76


## Step 3
Finds the two stats required across all patients: average BMI and the busiest month.


**average_bmi:** Loops through every exam, calls each object's own calculate_bmi(), and averages the results.

In [35]:
avg = average_bmi(exams)
print(f"{avg:.2f}")

21.60



**busiest_month:** Creates a {month: count} dictionary , then uses max(month_counts, key=month_counts.get) to find the month with the highest count, not the highest month number, which is what a plain max(month_counts) would give.

In [36]:
month = busiest_month(exams)
print(month)

6


## Normal Case
**Expected:** average BMI ≈ 21.60, busiest month = 6 (June).

In [37]:
exams_normal = load_exams("patient_data.csv")
avg_normal = average_bmi(exams_normal)
month_normal = busiest_month(exams_normal)
print("Rows:", len(exams_normal))
print(f"Average BMI: {avg_normal:.2f}")
print("Busiest month:", month_normal)

Rows: 1000
Average BMI: 21.60
Busiest month: 6


## Special case (One Patient)
**Expected:** average_bmi on a one item list just returns that patient's own BMI, and busiest_month returns that patient's own month.

In [40]:
single_exam = PatientExam(1, "3/5/2020", "John Smith", 70, 1.75)
single_list = [single_exam]
print(single_exam.calculate_bmi())
print(average_bmi(single_list))
print(busiest_month(single_list))

22.857142857142858
22.857142857142858
3


## Error case (CSV row missing a value (no height))
 **Expected:** IndexError, since load_exams accesses values[4] by position, and a 4-value row has no index 4, only 0-3.

In [43]:
try:
    load_exams("test.csv")
except IndexError as e:
    print("Caught:", e)

Caught: list index out of range
